### ISSUE TITLE
[Nice to have] Support script execution

### ISSUE THREAD

**Issue by denismakogon** [ronf/asyncssh#58]
For now users are to run command on remote host, but what if users want to run scripts? In most cases it would require to "re-invent the wheel" from time to time.

**Comment by nchammas** [comment]
You mean run a local script remotely? You can do that easily by copying the file up and executing it. Are you asking for a convenience method to do that for you?

**Comment by denismakogon** [comment]
Hello @nchammas, yes, exactly. So, in that solution i've been working on requires script execution functionality, so, what i've been doing is doing copying or serializing scripts and passing set of command to remote host to create script from content, setup environment variable and then run script from its path.

It would be nice to see such functionality in asyncssh if its possible.

**Comment by denismakogon** [comment]
@ronf @nchammas also, i've been looking for functionality that would allow tailing, similar to 'tail -f'. I've seen that in SSHReader you have readline method, so if i would want to run, for example:
    sudo apt-get upgrade -qy

and i want to get output of that command line by line, would readline work like that.
Basically, it would be nice to have method 'readlines' similar to python class file that would return lines generator.

**Comment by redixin** [comment]
> it would be nice to have method 'readlines' similar to python class file that would return lines generator.

It would be nice, but it wont work if separated stdout/stderr needed.

Callbacks seems good solution. Something like this:

    def run(cmd, stdin, stdout_callback, stderr_callback):
      """
      :returns exit status:
      """

    status = run("/bin/bash -xe -s", open("/local/script.sh"), sys.stdout.write, sys.stderr.write)

**Comment by redixin** [comment]
@denismakogon 

> it would be nice to have method 'readlines' similar to python class file that would return lines generator.

It would be nice, but it wont work if separated stdout/stderr needed.

Callbacks seems good solution. Something like this:

    def run(cmd, stdin, stdout_callback, stderr_callback):
      """
      :returns exit status:
      """

    status = run("/bin/bash -xe -s", open("/local/script.sh"), sys.stdout.write, sys.stderr.write)

**Comment by ronf** [comment]
I'm actually working on an easier to use wrapper for opening SSH client sessions. At the moment, I'm looking to implement something similar to subprocess (or asyncio.subprocess), but I would also probably include something like the new run() method that was added there in Python 3.5. See more at:

https://docs.python.org/3/library/subprocess.html

There are still some details to work out around which types of files or file descriptors I'll be able to allow and how to create pipelines involving multiple SSH sessions/connections, but the basic idea would be to allow an "async with" that would create a new session and let your access stdin/stdout/stderr in a slightly simpler form that open_session() does today by automatically closing and waiting for the session to end when leaving the block and an even simpler version you can "await" on if you just want to run a process and wait for it to complete, redirecting stdin/stdout/stderr but not doing interactive input & output during the lifetime of the process. That would look very much like what you show above.

Regarding the idea of making SSHReader support line-based iteration, there wasn't a clean way to do that with Python 3.4's iterators because you had to return coroutines and individually yield from them inside the loop. However, with the new "async for" support in Python 3.5, I think it may be possible now to do this. I'll add that to my list of things to investigate.

**Comment by ronf** [comment]
I just checked in a change to the 'develop' branch that should allow SSHReader to be used as an async iterator in Python 3.5, returning input data a line at a time similar to what you can do with Python file objects. The syntax looks something like:

```python
    stdin, stdout, stderr = await conn.open_session('ls')
    async for line in stdout:
        # Do something with line
```

Note: I didn't implement a readlines() function here, as that actually reads the entire file and returns a list of strings when run on file objects. Also, asyncio.StreamReader() doesn't implement readlines().

**Comment by denismakogon** [comment]
I was doing something more aligned to 'tail -f' (at least i think it looks like that). 

    while not eof:
        line = await stdout.readline()
        if line:
            node.context.logger.info('{1}'.format(node.name, line))
        eof = stdout.at_eof()


I don't really enjoy the idea if loading full stdout because it may appear too big, in my solution i'm not aware of what user is doing in their scripts, but it may be regular software updates after adding new source repos, upgrades, etc.

**Comment by ronf** [comment]
Yes - the async iterator I added works the same way as your loop. It does not load the entire output up front. Instead, it returns one line at a time as soon as a complete line is available without waiting for the rest of the output, exiting the loop when it hits EOF. So, it should be possible to replace your loop with:

```python
    async for line in stdout:
        node.context.logger.info('{1}'.format(node.name, line))
```

My comment about reading the entire output was in reference to the readlines() function, not what I added here. I didn't implement that because you generally wouldn't want to wait for all of the output before processing any of it, as you said.

**Comment by redixin** [comment]
@ronf 

    stdin, stdout, stderr = await conn.open_session('ls')
    async for line in stdout:
        # Do something with line

Remote process may produce a lot of mixed stdout and stderr data. What about stderr here? Will it be discarded or saved somehow?

**Comment by ronf** [comment]
With the open_session() call, you'll get back separate SSHStreamReader objects for stdout and stderr. So, you could have one coroutine doing an "async for line in stdout" and another running in parallel doing an "async for line in stderr".

Alternately, if you use the create_session() call, you can pass in a session class with a data_received() method that will get called whenever there's output to either stdout or stderr, so you can have a single handler processing both sets of output. The arguments to data_received() will tell you which stream the output is destined to.

With the new API I'm working on, you'll have a choice to send stdout and stderr directly to local files if you like, or process them as above by getting back SSHStrreamReader objects. You'll also have options to send both stdout and stderr to a single interleaved output stream if you don't need to distinguish between the two, or to discard output to stdout, stderr, or both.

**Comment by ronf** [comment]
I just checked in a new "SSHClientProcess" class into the develop branch, and functions create_process() and run() on the SSHClientConnection class that leverage it. This API provides a wrapper around the previous stream API that is modeled after the Python subprocess module and provides an easy way to redirect stdin/stdout/stderr or run a process and automatically collect all output from stdout and stderr into strings without having to do your own I/O calls. It also provides an easy way to redirect stdout and stderr to a single combined stream.

Arguments to stdin/stdout/stderr can be files, pipes, fifos, sockets, and other character devices, including TTYs once a bug I submitted a fix for in the asyncio module is fixed. SSHReader and SSHWriter classes from other SSHClientProcess objects can also be passed in to create pipelines of multiple processes. For convenience, files can be passed in multiple forms as well, including a string filename, open file-like object, or open low-level file descriptor.

Here's an example of using this new API:

```python
    async with asyncssh.connect('localhost') as conn:
        result = await conn.run('ls abc')
        print(result.stdout, end='')
```

Here's an example of feeding stdin from a static string and sending stdout to a local file:

```python
    async with asyncssh.connect('localhost') as conn:
        await conn.run('tail -r', input='1\n2\n3\n', stdout='/tmp/stdout')
```

Here's an example of piping one remote process to another and printing the output from the second process:

```python
    async with asyncssh.connect('localhost') as conn:
        proc1 = await conn.create_process(r'echo "1\n2\n3"')
        proc2_result = await conn.run('tail -r', stdin=proc1.stdout)
        print(proc2_result.stdout, end='')
```

Finally, here's a variant where a local process is spawned and piped to a remote process:

```python
    async with asyncssh.connect('localhost') as conn:
        local_proc = subprocess.Popen(r'echo "1\n2\n3"', shell=True, stdout=subprocess.PIPE)
        remote_result = await conn.run('tail -r', stdin=local_proc.stdout)
        print(remote_result.stdout, end='')
```

If anyone has some time to check this out, I'd love to get feedback on it. I'll cut a 1.6 release of AsyncSSH which includes these changes in the next couple of weeks.
